### PR TITLE
Changes to dataloader

### PR DESCRIPTION
**Pull request by mihirkatare**
Implemented in-built and hybrid dataloader to load event data and weights from root files efficiently. Added Readme and other documentation.

### CODE REVIEW

File: data.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    def __init__(self, manager):
+        self.manager = manager
+
+    def __len__(self):
+        file = uproot.open(self.manager.args["output_paths"][0])
+        len = file[self.manager.args["tree"]].arrays(self.manager.args["weights"], library="pd").shape[0]
+        file.close()
+        return len
+
+    def getitemhelper(self, idx):
```
Comment (matthewfeickert) [inline data.py:39]:
```suggestion
    def get_item_helper(self, idx):
```

It is generally a good idea to use Pythonic naming schemes. In Python classes are CameCase and everything else is snake_case. I'd apply snake_case to all functions and methods.

File: data.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+# path_weights = "/data/deepmem_debug/tt_20evt_weights.root"
+
+# file = uproot.open(path_weights)
+# a1 = file["t"].arrays(["bb_p4/fCoordinates/fCoordinates.fPt", "bb_p4/fCoordinates/fCoordinates.fEta", "bb_p4/fCoordinates/fCoordinates.fPhi", "bb_p4/fCoordinates/fCoordinates.fM"], library="pd").values.T
+# print(a1)
+# print(file["t"].show())
+
+# file.close()
+
+if(__name__ == "__main__"):
```
Comment (matthewfeickert) [inline data.py:89]:
```suggestion
if __name__ == "__main__":
```

The `()` aren't needed and don't do anything here.

File: data.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+# path = "/data/deepmem_debug/tt_20evt.root"
+# path_weights = "/data/deepmem_debug/tt_20evt_weights.root"
+
+# file = uproot.open(path_weights)
+# a1 = file["t"].arrays(["bb_p4/fCoordinates/fCoordinates.fPt", "bb_p4/fCoordinates/fCoordinates.fEta", "bb_p4/fCoordinates/fCoordinates.fPhi", "bb_p4/fCoordinates/fCoordinates.fM"], library="pd").values.T
+# print(a1)
+# print(file["t"].show())
+
+# file.close()
```
Comment (matthewfeickert) [inline data.py:87]:
```suggestion
```

Commented out code is better to have removed.

File: data.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+if(__name__ == "__main__"):
+    # This script is for testing the runtime of the dataloader.
+    # It has a similar structure to a common PyTorch training loop.
```
Comment (matthewfeickert) [inline data.py:91]:
```suggestion
    """
    This script is for testing the runtime of the dataloader.
    It has a similar structure to a common PyTorch training loop.
    """
```

This is a good use of triple quote docstring style comments.

File: dataloaders/data_hybrid.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
+import sys
+sys.path.append('./') # append the root directory
```
Comment (matthewfeickert) [inline dataloaders/data_hybrid.py:2]: We should make this more robust later, but for now it is okay.

File: dataloaders/data_hybrid.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+        file = uproot.open(self.manager.args["output_paths"][0])
+        self.len = file[self.manager.args["tree"]].num_entries # Total number of samples
+        file.close()
+
+        self.n_chunks = self.len // self.nec
+
+    def __len__(self):
+        return self.n_chunks*self.bic
+
+    def loadChunk(self, n):
```
Comment (matthewfeickert) [inline dataloaders/data_hybrid.py:54]: Same comments above RE: snake_case apply to this file.

File: dataloaders/data_hybrid.py [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+        # Resolve relative paths to absolute paths
+        self.args["input_paths"] = list(map( (lambda x : Path(x).resolve()), self.args["input_paths"]))
+        self.args["output_paths"] = list(map( (lambda x : Path(x).resolve()), self.args["output_paths"]))
+
+    def string_processing(self, string):
+        return( string + "/fCoordinates/fCoordinates." )
+
+class CustomDSHybrid(Dataset):
+    # Implementation using a modified PyTorch dataset but DOES NOT support multithreading (i.e. workers=0 must be used).
```
Comment (matthewfeickert) [inline dataloaders/data_hybrid.py:36]: Would use docstring quotes here

File: dataloaders/data_hybrid.py [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+            self.loadChunk(idx // self.nec)
+
+        if(chunkIdx == self.bic-1):
+            return self.X_chunk[chunkIdx*self.bs:], self.Y_chunk[chunkIdx*self.bs:]
+        else:
+            return self.X_chunk[chunkIdx*self.bs:(chunkIdx+1)*self.bs], self.Y_chunk[chunkIdx*self.bs:(chunkIdx+1)*self.bs]
+
+
+class CustomDSHybridMultithread(Dataset):
+    # Implemntation of the above class but in a way that multithreading can be used. However this cannot be used
```
Comment (matthewfeickert) [inline dataloaders/data_hybrid.py:86]: Would use docstring quotes here

File: dataloaders/data_hybrid.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+
+# path = "/data/deepmem_debug/tt_20evt.root"
+# path_weights = "/data/deepmem_debug/tt_20evt_weights.root"
+
+# file = uproot.open(path_weights)
+# a1 = file["t"].arrays(["bb_p4/fCoordinates/fCoordinates.fPt", "bb_p4/fCoordinates/fCoordinates.fEta", "bb_p4/fCoordinates/fCoordinates.fPhi", "bb_p4/fCoordinates/fCoordinates.fM"], library="pd").values.T
+# print(a1)
+# print(file["t"].show())
+
+# file.close()
```
Comment (matthewfeickert) [inline dataloaders/data_hybrid.py:145]:
```suggestion
```

File: dataloaders/data_hybrid.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+if(__name__ == "__main__"):
```
Comment (matthewfeickert) [inline dataloaders/data_hybrid.py:147]:
```suggestion
if__name__ == "__main__":
```

File: dataloaders/data_hybrid.py [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+    # file.close()
+    # n_chunks = len // data_manager.args["chunk_entries"]
+    # for epoch in range(1):
+    #     for chunk in range(n_chunks):
+    #         dataset = CustomDSHybridMultithread(data_manager, whichChunk=chunk)
+    #         generator = DataLoader(dataset, batch_size=1, shuffle=False, num_workers = 3)
+
+    #         for batch_idx, data in enumerate(generator):
+    #             x, y = data
+    #             print(x.shape, y.shape, " -> batch ", batch_idx)
```
Comment (matthewfeickert) [inline dataloaders/data_hybrid.py:176]:
```suggestion
```

If there are things that you want to keep around to return to, I'd just keep them on another branch but I'd suggest not adding commented out code into `main`.

**Review by matthewfeickert — COMMENTED** [review]
I realize that I didn't talk about Black and flake8 in #1. After this PR goes in we should talk about those tools and I can open a PR that adds them to pre-commit.

File: data.py (hunk 1 above)
Comment (mihirkatare) [reply data.py:39]: Hello, while I can change "getitemhelper" to "get_item_helper" the other functions like "__getitem__" (for which getitemhelper is a helper method) are inherited from the base Dataset class and cannot be changed.

Comment (matthewfeickert) [reply data.py:39]: Yeah not much you can do there. In that case it is probably better to have a more consistent API.

**Comment by matthewfeickert** [comment]
Sorry for the long delay on this @mihirkatare &mdash; I hadn't seen a re-request for review so I hadn't checked for a bit. But all LGTM. :+1:

**Pull request merged**